src/connection.ts

```diff
@@ -182,6 +197,74 @@ export class ConnectionManager implements IConnectionManager {
     callback(proc);
   }
 
+  private createUniqueDockerName() {
+    return 'puppet-vscode-xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
+      const r = Math.random() * 16 | 0, v = c === 'x' ? r : ( r & 0x3 | 0x8 );
+      return v.toString(16);
+    });
+  }
+
+  private getDockerPort(name: string) {
+    let cmd: string = 'docker.exe';
+    let args: Array<string> = [
+      "port",
+      this.dockerName,
+      '8082'
+    ];
+    var proc = cp.spawnSync(cmd, args);
+    // docker port {id} {port} returns output like 0.0.0.0:1234
+    // We want to retrieve that port for use in the TCP connection.
+    let regex = /:(\d+)$/m;
+    return Number(regex.exec(proc.stdout.toString())[1]);
+  }
+
+  public startLanguageServerDockerProcess( options: cp.SpawnOptions, callback: Function) {
+    let logPrefix: string = '[startLanguageServerProcessAsContainer] ';
+    this.dockerName = this.createUniqueDockerName();
+    // TODO: get settings to specify image name and tag
+    let cmd: string = 'docker.exe';
+    let args: Array<string> = [
+      "run",
+      '--rm',
+      '-i',
+      '-P',
+      '--name',
+      this.dockerName,
+      `${this.connectionConfiguration.dockerImageName}:${this.connectionConfiguration.dockerImageTag}`
+    ];
+    console.log(vscode.CompletionItemKind.Class);
+    // Is this windows only? When run on windows PATH is just the ruby paths, Path has everything else.
+    options.env.PATH = process.env.PATH;
+    this.logger.debug(logPrefix + 'Starting the language server as a container with ' + cmd + ' ' + args.join(' '));
+    var proc = cp.spawn(cmd, args, options);
+    this.logger.debug(logPrefix + 'Language server PID:' + proc.pid);
+    callback(proc);
+  }
+
+  // TODO: Should this return anything? /shrug
+  private stopLanguageServerDockerProcess(): void {
+    let logPrefix: string = '[stopLanguageServerProcessAsContainer] ';
+    let cmd: string = 'docker.exe';
+    let args: Array<string> = [
+      "rm",
+      '--force',
+      this.dockerName,
+    ];
+    //TODO: Is this necessary?
```

Comment:
Wasn't sure if the spawn_options are actually needed. If not, we can clean out some duplicate code.